File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    static bool compare(const vector<int>& interval1, const vector<int>& interval2) {
        return interval1[1] < interval2[0];
    }
    
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        auto range = equal_range(intervals.begin(), intervals.end(), newInterval, compare);
        auto itr1 = range.first, itr2 = range.second;
        if (itr1 == itr2) {
            intervals.insert(itr1, newInterval);
        } else {
            itr2--;
            *(itr2->begin() + 0) = min(newInterval[0], *(itr1->begin() + 0));
            *(itr2->begin() + 1) = max(newInterval[1], *(itr2->begin() + 1));
            intervals.erase(itr1, itr2);
        }
        return intervals;
    }
};

File: 0057-insert-interval/0057-insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        ArrayList<int[]> mergedIntervals = new ArrayList<>(); // ArrayList to store merged intervals
        for (int[] interval : intervals) {
            // If the end of the new interval is before the start of the current interval
            if (newInterval[1] < interval[0]) {
                mergedIntervals.add(newInterval); // Add the new interval to the result
                newInterval = interval; // Update new interval to current interval for future comparisons
            } 
            // If the end of the current interval is before the start of the new interval
            else if (interval[1] < newInterval[0]) {
                mergedIntervals.add(interval); // Add the current interval to the result
            } 
            // If there is an overlap between the new interval and the current interval
            else {
                // Merge the intervals by updating the start and end of the new interval
                newInterval[0] = Math.min(newInterval[0], interval[0]); // Update start of new interval
                newInterval[1] = Math.max(newInterval[1], interval[1]); // Update end of new interval
            }
        }
        mergedIntervals.add(newInterval); // Add the final merged interval
        return mergedIntervals.toArray(new int[mergedIntervals.size()][]); // Convert ArrayList to array
    }
}